<a href="https://colab.research.google.com/github/Abdelrhman129/my-_colab_notbook/blob/main/Data_Anlalysis_Phython.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Analysis with Python: Part 1 of 6

**Part_one(frist_video)**

In [ ]:
2+3+9

14

In [ ]:
99-73

26

In [ ]:
23.54 * -1432

-33709.28

In [ ]:
100/7

14.285714285714286

In [ ]:
100//7

14

In [ ]:
100 % 7

2

In [ ]:
5 ** 3

125

In [ ]:
((2+5)*(17-3))/(4**3)

1.53125

Python supports the following arithmetic operators:

| Operator   | Purpose           | Example     | Result    |
|------------|-------------------|-------------|-----------|
| `+`        | Addition          | `2 + 3`     | `5`       |
| `-`        | Subtraction       | `3 - 2`     | `1`       |
| `*`        | Multiplication    | `8 * 12`    | `96`      |
| `/`        | Division          | `100 / 7`   | `14.28..` |
| `//`       | Floor Division    | `100 // 7`  | `14`      |    
| `%`        | Modulus/Remainder | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |


Try solving some simple problems from this page: https://www.math-only-math.com/worksheet-on-word-problems-on-four-operations.html . You can use the empty cells below and add more cells if required.

[ ]
Solving multi-step problems using variables
Let's try solving the following word problem using Python:

A grocery store sells a bag of ice for $1.25, and makes 20% profit. If it sells 500 bags of ice, how much total profit does it make?

We can list out the information provided, and gradually convert the word problem into a mathematical expression which can be evaluated using Python.

Cost of ice bag ($) = 1.25

Profit margin = 20% = .2

Profit per bag ($) = profit margin * cost of ice bag = .2 * 1.25

No. of bags = 500

Total profit = no. of bags * profit per bag = 500 * (.2 * 1.25)

In [ ]:
500*(1.25*2)

1250.0

In [ ]:
cost_of_ice_bag = 1.25

In [ ]:
profit_margin = .2

In [ ]:
number_of_bags = 500

In [ ]:
profit_margin

0.2

In [ ]:
profit_per_bag = cost_of_ice_bag * profit_margin

In [ ]:
total_profit = number_of_bags * profit_per_bag

In [ ]:
total_profit

125.0

In [ ]:
print("The grocery store makes a total profit of $",total_profit)

The grocery store makes a total profit of $ 125.0


In [ ]:
# Store input data in variables
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


Evaluating conditions using Python
Apart from arithmetic operations, Python also provides several oprations for comparing numbers & variables.

Operator	Description
==	Check if operands are equal
!=	Check if operands are not equal
>	Check if left operand is greater than right operand
<	Check if left operand is less than right operand
>=	Check if left operand is greater than or equal to right operand
<=	Check if left operand is less than or equal to right operand
The result of a comparision operation is either True or False (note the uppercase T and F). These are special keywords in Python. Let's try out some experiment with comparision operators.

In [ ]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [ ]:
# Equality check_True
my_favorite_number == 1

True

In [ ]:
# Equality check - False
my_favorite_number == my_least_favorite_number

False

In [ ]:
# Not equal check - True
my_favorite_number != a_neutral_number

True

In [ ]:
# Not equal check - False
a_neutral_number != 3

False

In [ ]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

True

In [ ]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >=10
print("Is the ice bag expensive?", is_ice_bag_expensive)

Is the ice bag expensive? False


## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evalute to `True`. Otherwise it returns `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |


In [ ]:
my_favorite_number

1

In [ ]:
my_favorite_number >0 and my_favorite_number <=3

True

In [ ]:
my_favorite_number <0 and my_favorite_number <=3

False

In [ ]:
my_favorite_number >0 and my_favorite_number >=3

False

In [ ]:
True and False

False

In [ ]:
True and True

True

The `or` operator returns `True` if at least one of the conditions evalute to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |


In [ ]:
a_neutral_number

3

In [ ]:
a_neutral_number == 3 or my_favorite_number < 0

True

In [ ]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [ ]:
my_favorite_number < 0 or True

True

In [ ]:
False or False

False

In [ ]:
not a_neutral_number ==3

False

In [ ]:
not False

True

In [ ]:
not True

False

Logical operators can be combined to form complex conditions. Use round brackets or parantheses ( and ) to indicate the order in which logical operators should be applied.

In [ ]:
(2>3 and 4<=5) or not (my_favorite_number <0 and True)

True

In [ ]:
not (True and 0 < 1) or (False and True)

False

In [ ]:
not True and 0 < 1 or False and True

False

# A Quick Tour of Variables and Data Types in Python

![](https://i.imgur.com/6cg2E9Q.png)


This tutorial series is a beginner-friendly introduction to programming and data analysis using the Python programming language. These tutorials take a practical and coding-focused approach. The best way to learn the material is to execute the code and experiment with it yourself.

In [ ]:
my_favorite_color="blue"

In [ ]:
my_favorite_color

'blue'

In [ ]:
color1,color2,color3 ="red","blue","Black"

In [ ]:
color1

'red'

In [ ]:
color2

'blue'

In [ ]:
counter=10

In [ ]:
counter+5

15

In [ ]:
counter=10

In [ ]:

counter+=3


In [ ]:
counter

13

In [ ]:
a_variable=21
is_today_suterday=False
my_favorite_car="Delorean"
the_3_musketeers=["Athos","porthos","Armais"]

In [ ]:
type(a_variable)

int

In [ ]:
type(is_today_suterday)

bool

Python has several built-in data types for storing different kinds of information in variables. Following are some commonly used data types:

1. Integer
2. Float
3. Boolean
4. None
5. String
6. List
7. Tuple
8. Dictionary

Integer, float, boolean, None, and string are *primitive data types* because they represent a single value. Other data types like list, tuple, and dictionary are often called *data structures* or *containers* because they hold multiple pieces of data together.

In [ ]:
str(23)

'23'

In [ ]:
fruits = ['apple', 'banana', 'cherry']

In [ ]:
print("the number of fruits",len(fruits))

the number of fruits 3


In [ ]:
fruits.append('dates')

In [ ]:
fruits

['apple', 'banana', 'cherry', 'dates']

In [ ]:
fruits[1]="bllueberry"

In [ ]:
fruits

['apple', 'bllueberry', 'cherry', 'dates']

In [ ]:
fruits.insert(1,'banana')

In [ ]:
fruits

['apple', 'banana', 'bllueberry', 'cherry', 'dates']

In [ ]:
fruits.pop(1)

'banana'

In [ ]:
fruits

['apple', 'bllueberry', 'cherry', 'dates']

In [ ]:
fruits.pop()

'dates'

In [ ]:
fruits

['apple', 'bllueberry', 'cherry']

### Dictionary

A dictionary is an unordered collection of items. Each item stored in a dictionary has a key and value. You can use a key to retrieve the corresponding value from the dictionary.  Dictionaries have the type `dict`.

Dictionaries are often used to store many pieces of information e.g. details about a person, in a single variable. Dictionaries are created by enclosing key-value pairs within braces or curly brackets `{` and `}`.

In [ ]:
personl={
    "name":'Abdelrahman',
    'sex':'Male',
    'age':'24',
    'married':True

}

In [ ]:
personl

{'name': 'Abdelrahman', 'sex': 'Male', 'age': '24', 'married': True}

In [ ]:
personl2=dict(name='Ahmed',sex='Male',age='12',married=False)

In [ ]:
personl2

{'name': 'Ahmed', 'sex': 'Male', 'age': '12', 'married': False}

In [ ]:
type(personl)

dict

In [ ]:
personl["name"]


'Abdelrahman'

In [ ]:
"addries" in personl2

False

In [ ]:
"name" in personl2

True

# NUMPY

In [ ]:
w1 ,w2 ,w3 =0.3,0.2,0.5

In [ ]:
kanto_temp=73
kanto_rainfall=67
kanto_humidity=43

In [ ]:
kanto_yeild_apples=kanto_temp*w1+kanto_rainfall*w2+kanto_humidity*w3

In [ ]:
print("the expected yeild of apple in kanto region is {} tons per hectare".format(kanto_yeild_apples))

the expected yeild of apple in kanto region is 56.8 tons per hectare


In [ ]:
kanto=[73,67,43]
johto=[91,88,64]
hoenn=[87,134,58]
sinnoh=[102,43,37]
unova=[69,96,70]

In [ ]:
wieghts=[w1,w2,w3]

In [ ]:
def crop_yield(region,wieghts):
  result=0
  for x,w in zip(region,wieghts):
    result+=x*w
  return result


In [ ]:
crop_yield(kanto,wieghts)

56.8

In [ ]:
crop_yield(johto,wieghts)

76.9

In [ ]:
crop_yield(hoenn,wieghts)

81.9

In [ ]:
crop_yield(sinnoh,wieghts)

57.699999999999996

In [ ]:
import numpy as np

In [ ]:
kanto=np.array([79,67,43])

In [ ]:
kanto

array([79, 67, 43])

In [ ]:
wieghts=np.array([w1,w2,w3])

In [ ]:
wieghts

array([0.3, 0.2, 0.5])

In [ ]:
type(kanto)

numpy.ndarray

In [ ]:
np.dot(kanto,wieghts)

np.float64(58.6)

In [ ]:
(kanto*wieghts).sum()

np.float64(58.6)

In [ ]:
arr1=np.array([1,2,3])
arr2=np.array([4,5,6])


In [ ]:
arr1*arr2

array([ 4, 10, 18])

In [ ]:
arr2.sum()

np.int64(15)

In [ ]:
wieghts

array([0.3, 0.2, 0.5])

In [ ]:
climate_data

NameError: name 'climate_data' is not defined

In [ ]:
climate_data=np.array([[73,67,43],
                      [91,88,64],
                      [87,134,58],
                      [102,43,37],
                      [69,96,70]])








In [ ]:
climate_data

In [ ]:
climate_data.shape

In [ ]:
type(climate_data)

In [ ]:
climate_data.shape

In [ ]:
wieghts.shape

In [ ]:
arr3=np.array([
    [[11,12,13],
     [13,14,15]],
    [[15,16,17],
     [17,18,19.5]]])

In [ ]:
arr3.shape

In [ ]:
print(arr3.shape)

In [ ]:
climate_data.dtype

In [ ]:
np.dot(climate_data,wieghts)

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("countries of the world.csv")
df

In [ ]:
df = pd.read_csv("countries of the world.csv",header=None)
df

In [ ]:
df = pd.read_csv("countries of the world.csv",header=None,names=['country','region'])
df


In [ ]:
df=pd.read_csv("countries of the world (2).txt")
df

In [ ]:
df=pd.read_table("countries of the world (2).txt",sep='\t')

df

In [ ]:
df=pd.read_json("json_sample.json")
df

In [ ]:
df2=pd.read_excel("world_population_excel_workbook.xlsx",sheet_name='world_population')
df2

In [ ]:
df2.shape

In [ ]:
pd.set_option('display.max.rows',236)
pd.set_option('display.max.columns',40)

In [ ]:
df2.info()

In [ ]:
pd.reset_option('display.max.rows')
pd.reset_option('display.max.columns')

In [ ]:
df2.head()

In [ ]:
df2.tail()

In [ ]:
df2['Rank']

In [ ]:
df2['Continent']

In [ ]:
df2.loc[233]

NameError: name 'df2' is not defined

In [ ]:
df2.iloc[233]

In [ ]:
df3=pd.read_csv("world_population.csv")
df3

In [ ]:
df3[df3['Rank']<10]

In [ ]:
specific_countries=['Bangladesh','Brazil']
df3[df3['Country'].isin(specific_countries)]

In [ ]:
df3[df3['Country'].str.contains('Brazil')]

In [ ]:
df=df3.set_index('Country')
df

In [ ]:
df.filter(items=['Continent','CCA3'],axis=1)

In [ ]:
df.filter(items=['Zimbabwe'],axis=0)

In [ ]:
df.filter(like='United',axis=0)

In [ ]:
df.loc['United States']

In [ ]:
df.iloc[3]

,1004
First_Name,Dwight
Last_Name,Schrute
Phone_Number,1235432345
Address,"980 Paper Avenue, Pennsylvania, 18503"
Paying Customer,Yes
Do_Not_Contact,Y
Not_Useful_Column,True


In [ ]:
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,1235455421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,1236439775,93 West Main Street,No,Yes,False
1003,Walter,White,,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,1235432345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,8766783469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,3047622467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,,1209 South Street,No,No,False
1008,Sherlock,Holmes,8766783469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,,123 Middle Earth,Yes,NaN,False


In [ ]:
df[df['Rank']<10].sort_values(by='Rank',ascending=True)

KeyError: 'Rank'

In [ ]:
 df[df['Rank']<10].sort_values(by=['Continent','Country'],ascending=False)

In [ ]:
 df[df['Rank']<10].sort_values(by=['Continent','Country'],ascending=[False,True])

In [ ]:
import pandas as pd

In [ ]:
df=pd.read_csv("Flavors.csv")
df

In [ ]:
group_by_frame = df.groupby('Base Flavor')

In [ ]:
group_by_frame.mean(numeric_only=True)

In [ ]:
df.groupby('Base Flavor').count()

In [ ]:
df.groupby('Base Flavor').min()

In [ ]:
df.groupby('Base Flavor').max()

In [ ]:
df.groupby('Base Flavor').sum(numeric_only=True)

KeyError: 'Base Flavor'

In [ ]:
df.groupby("Base Flavor").agg({'Flavor Rating':['mean','max','min','count','sum']})

In [ ]:
df.groupby("Base Flavor").agg({'Flavor Rating':['mean','max','min','count','sum'],'Total Rating':['mean','max','min','count','sum']})
#

In [ ]:
df.groupby(['Base Flavor','Liked']).mean(numeric_only=True)

In [ ]:
df.groupby(['Base Flavor','Liked']).agg({'Flavor Rating':['mean','max','min','count','sum']})
#

In [ ]:
df.groupby("Base Flavor").describe()

**🧠 أهم حاجة تتعلمها هنا
1️⃣ مش بس تبص على mean ❌

لازم تبص على:

min / max
quartiles**

**2️⃣ الفرق بين 25% و 75%

👉 لو الفرق صغير → الداتا ثابتة
👉 لو كبير → فيه تشتت**

In [ ]:
df2=pd.read_csv("LOTR 2.csv")
df2

In [ ]:
df1=pd.read_csv("LOTR.csv")
df1

FileNotFoundError: [Errno 2] No such file or directory: 'LOTR.csv'

In [ ]:
df1.merge(df2)

In [ ]:
df1.merge(df2,how='inner')

In [ ]:
df1.merge(df2,how='inner',on='FellowshipID')

NameError: name 'df1' is not defined

In [ ]:
df1.merge(df2,how='inner',on='FirstName')

In [ ]:
df1.merge(df2,how='inner',on=['FirstName','FellowshipID'])

In [ ]:
df1.merge(df2,how='outer')

NameError: name 'df1' is not defined

In [ ]:
df1.merge(df2,how='left')

In [ ]:
df1.merge(df2,how='right')

In [ ]:
df1.merge(df2,how='cross')

NameError: name 'df1' is not defined

In [ ]:
df1.join(df2,on='FellowshipID',how='outer',lsuffix='_Left',rsuffix='_Right')

In [ ]:
df4 = df1.set_index('FellowshipID').join(df2.set_index('FellowshipID'), lsuffix = '_Left',rsuffix = '_Right', how = 'outer')
df4

In [ ]:
pd.concat([df1,df2], join = 'outer', axis = 1)


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [ ]:
df=pd.read_csv("Ice Cream Ratings (1).csv")
df

In [ ]:
df=df.set_index('Date')
df

In [ ]:
df.groupby('Date').describe()

In [ ]:
df.plot ()

TypeError: no numeric data to plot

In [ ]:
df.plot(kind='bar',stacked=True)

In [ ]:
df.plot(kind='line',title='Ice Cream Ratings',grid=True,xlabel="Daily Rating",ylabel="Rating")


In [ ]:
df.plot(kind='hist')


In [ ]:
df['Flavor Rating'].plot(kind='bar',stacked=True)

In [ ]:
df.boxplot()

In [ ]:
df.plot.area()

In [ ]:
df.plot.bar()

In [ ]:
df.plot.pie(y='Flavor Rating',figsize=(10,6))

In [2]:
import pandas as pd

In [3]:
df=pd.read_excel("Customer Call List.xlsx")
df

,CustomerID,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
0,1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1,1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes,False
2,1003,Walter,/White,7066950392,298 Drugs Driveway,N,NaN,True
3,1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
4,1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No,True
5,1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
6,1007,Jeff,Winger,NaN,1209 South Street,No,No,False
7,1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No,False
8,1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN,False
9,1010,Peter,Parker,123-545-5421,"25th Main Street, New York",Yes,No,True


In [4]:
pd.set_option('display.max.columns',200)
pd.set_option("display.max.rows",200)

In [5]:
df=df.set_index('CustomerID')
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes,False
1003,Walter,/White,7066950392,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN,False


In [6]:
df.drop_duplicates()

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes,False
1003,Walter,/White,7066950392,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN,False


In [7]:
df.drop(columns='Not_Useful_Column')

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact
CustomerID,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No
1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes
1003,Walter,/White,7066950392,298 Drugs Driveway,N,NaN
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y
1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes
1007,Jeff,Winger,NaN,1209 South Street,No,No
1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No
1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN


In [8]:
df['Last_Name']=df['Last_Name'].str.lstrip('/')
df['Last_Name']=df['Last_Name'].str.rstrip('_')
df['Last_Name']=df['Last_Name'].str.rstrip('...')


df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes,False
1003,Walter,White,7066950392,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN,False


In [9]:

df['Last_Name']=df['Last_Name'].str.strip('123._/')
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes,False
1003,Walter,White,7066950392,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN,False


In [ ]:
df['Last_Name']

,Last_Name
CustomerID,
1001,Baggins
1002,Nadir
1003,White
1004,Schrute
1005,Snow
1006,Swanson
1007,Winger
1008,Holmes
1009,NaN


In [ ]:
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes,False
1003,Walter,White,7066950392,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN,False


In [10]:
df['Phone_Number'].str.replace("[^a-zA-z0-9]",' ')


,Phone_Number
CustomerID,
1001,123-545-5421
1002,123/643/9775
1003,NaN
1004,123-543-2345
1005,876|678|3469
1006,304-762-2467
1007,NaN
1008,876|678|3469
1009,N/a


In [13]:
df['Phone_Number']=df['Phone_Number'].str.replace("[^a-zA-Z0-9]", '', regex=True)
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,1235455421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,1236439775,93 West Main Street,No,Yes,False
1003,Walter,White,NaN,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,1235432345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,8766783469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,3047622467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,8766783469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,Na,123 Middle Earth,Yes,NaN,False


In [11]:
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,123/643/9775,93 West Main Street,No,Yes,False
1003,Walter,White,7066950392,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,876|678|3469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,876|678|3469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,N/a,123 Middle Earth,Yes,NaN,False


In [12]:
import numpy as np

df['Phone_Number'] = df['Phone_Number'].astype(str)

# نشيل أي حاجة غير الأرقام
df['Phone_Number'] = df['Phone_Number'].str.replace(r'[^0-9]', '', regex=True)

# نخلي الرقم اللي طوله مش 10 يبقى NaN
df['Phone_Number'] = df['Phone_Number'].apply(lambda x: x if len(x)==10 else np.nan)

# نعمل فورمات 123-456-7890
df['Phone_Number'] = df['Phone_Number'].apply(lambda x: x[0:3]+'-'+x[3:6]+'-'+x[6:10] if pd.notna(x) else np.nan)

In [13]:
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column
CustomerID,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True
1002,Abed,Nadir,123-643-9775,93 West Main Street,No,Yes,False
1003,Walter,White,706-695-0392,298 Drugs Driveway,N,NaN,True
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True
1005,Jon,Snow,876-678-3469,123 Dragons Road,Y,No,True
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True
1007,Jeff,Winger,NaN,1209 South Street,No,No,False
1008,Sherlock,Holmes,876-678-3469,98 Clue Drive,N,No,False
1009,Gandalf,NaN,NaN,123 Middle Earth,Yes,NaN,False


In [16]:
df['Phone_Number'].str.replace('NaN','')

,Phone_Number
CustomerID,
1001,123-545-5421
1002,123-643-9775
1003,NaN
1004,123-543-2345
1005,876-678-3469
1006,304-762-2467
1007,NaN
1008,876-678-3469
1009,NaN


In [14]:
df['Phone_Number']=df['Phone_Number'].fillna('')
df

In [19]:
df['Address']

,Address
CustomerID,
1001,"123 Shire Lane, Shire"
1002,93 West Main Street
1003,298 Drugs Driveway
1004,"980 Paper Avenue, Pennsylvania, 18503"
1005,123 Dragons Road
1006,768 City Parkway
1007,1209 South Street
1008,98 Clue Drive
1009,123 Middle Earth


In [23]:
df['Address'].str.split(',',n=1,expand=True)


,0,1
CustomerID,,
1001,123 Shire Lane,Shire
1002,93 West Main Street,None
1003,298 Drugs Driveway,None
1004,980 Paper Avenue,"Pennsylvania, 18503"
1005,123 Dragons Road,None
1006,768 City Parkway,None
1007,1209 South Street,None
1008,98 Clue Drive,None
1009,123 Middle Earth,None


In [21]:
df['Address'].str.split(pat=',', n=2, expand=True)

,0,1,2
CustomerID,,,
1001,123 Shire Lane,Shire,None
1002,93 West Main Street,None,None
1003,298 Drugs Driveway,None,None
1004,980 Paper Avenue,Pennsylvania,18503
1005,123 Dragons Road,None,None
1006,768 City Parkway,None,None
1007,1209 South Street,None,None
1008,98 Clue Drive,None,None
1009,123 Middle Earth,None,None


In [25]:
df[['street_Address','state','Zip_code']]=df['Address'].str.split(pat=',', n=2, expand=True)
df

,First_Name,Last_Name,Phone_Number,Address,Paying Customer,Do_Not_Contact,Not_Useful_Column,street_Address,state,Zip_code
CustomerID,,,,,,,,,,
1001,Frodo,Baggins,123-545-5421,"123 Shire Lane, Shire",Yes,No,True,123 Shire Lane,Shire,None
1002,Abed,Nadir,123-643-9775,93 West Main Street,No,Yes,False,93 West Main Street,None,None
1003,Walter,White,706-695-0392,298 Drugs Driveway,N,NaN,True,298 Drugs Driveway,None,None
1004,Dwight,Schrute,123-543-2345,"980 Paper Avenue, Pennsylvania, 18503",Yes,Y,True,980 Paper Avenue,Pennsylvania,18503
1005,Jon,Snow,876-678-3469,123 Dragons Road,Y,No,True,123 Dragons Road,None,None
1006,Ron,Swanson,304-762-2467,768 City Parkway,Yes,Yes,True,768 City Parkway,None,None
1007,Jeff,Winger,,1209 South Street,No,No,False,1209 South Street,None,None
1008,Sherlock,Holmes,876-678-3469,98 Clue Drive,N,No,False,98 Clue Drive,None,None
1009,Gandalf,NaN,,123 Middle Earth,Yes,NaN,False,123 Middle Earth,None,None


In [27]:
df['Paying Customer'].set.replace('Y','Yes',regex=True)

AttributeError: 'Series' object has no attribute 'set'